Support `@layer` (#955)
vanilla-extract-css/vanilla-extract: site/docs/api/layer.md

```

## Nesting layers

To facilitate organisation of styles, [layer nesting] is supported by providing a `parent` layer reference via the options object.
This will generate the shorthand syntax, i.e. `parent.child`, while also making the relationship between layers explicit.

```ts compiled
// layers.css.ts

import { layer } from '@vanilla-extract/css';

export const reset = layer('reset');
export const framework = layer('framework');
export const typography = layer(
  { parent: framework },
  'typography'
);
```

## Assigning styles

Styles can be assigned to a layer by name, using the `@layer` key in the style object.

In this example, we first import the `layers.css.ts` stylesheet, setting up the order of the layers, then create a style within the `reset` layer.

```ts compiled
// reset.css.ts
import { style } from '@vanilla-extract/css';
import { reset } from './layers.css.ts';

export const noMargin = style({
  '@layer': {
    [reset]: {
      margin: 0
    }
  }
});

// layers.css.ts
import { layer } from '@vanilla-extract/css';

export const reset = layer('reset');
export const framework = layer('framework');
export const typography = layer(

>   { parent: 'framework' },

Reviewer: ```suggestion
  { parent: framework },
```

  'typography'
);
```

## Layer merging

> In order to generate the smallest possible CSS output, Vanilla Extract will merge styles that are assigned to the same layer, if it can be done without impacting the precedence of the rules.

Reviewer 2: ```suggestion
In order to generate the smallest possible CSS output, Vanilla Extract will merge styles that are assigned to the same layer within the same file, if it can be done without impacting the precedence of the rules.
```

> Notice in this example, while the `themedHeading` style is created before the other styles, it appears later in the stylesheet. This is due to it being assigned to the `theme` layer — which is declared after the `base` layer.
> 
> ```ts compiled
> // typography.css.ts
> import { style, layer } from '@vanilla-extract/css';
> 
> const base = layer();
> const theme = layer();
> 
> const themedHeading = style({
>   '@layer': {
>     [theme]: {
>       color: 'rebeccapurple'
>     }
>   }
> });
> const text = style({
>   '@layer': {
>     [base]: {
>       fontSize: '1rem'
>     }
>   }
> });

Author: Tweaked wording slightly to be in line with the latest implementation.

```suggestion
Notice in this example, while the `themedHeading` style is created before the the `heading` style, it appears later in the stylesheet. This is due to it being assigned to the `theme` layer — which is declared after the `base` layer.

```ts compiled
// typography.css.ts
import { style, layer } from '@vanilla-extract/css';

const base = layer();
const theme = layer();

const text = style({
  '@layer': {
    [base]: {
      fontSize: '1rem'
    }
  }
});
const themedHeading = style({
  '@layer': {
    [theme]: {
      color: 'rebeccapurple'
    }
  }
});
```

The new output looks like this:

```css
@layer base;
@layer theme;
@layer base {
  .typography_text__94ldqt0 {
    font-size: 1rem;
  }
  .typography_heading__94ldqt2 {
    font-size: 2.4rem;
  }
}
@layer theme {
  .typography_themedHeading__94ldqt1 {
    color: rebeccapurple;
  }
}
```
